### DISCUSSION TITLE
JSON.stringify

### DISCUSSION THREAD

**Discussion by marchbold** [airsdk/Adobe-Runtime-Support#920] [Q&A] [answered]
Does anyone know if there is a way to control the order of the content written from `JSON.stringify`?

I understand that json isn't ordered, and it doesn't technically matter, however I was hoping to have a little control over what is output for appearances sake. 

---

This is for some work I'm doing on APM.

```json
{
    "id": "com.my.package",
    "name": "My Package",
    "url": "https://example.com/package/",
    "description": "An example of a package definition",
    "type": "swc",
    "version": "1.0.0"  
}
```

I am hoping to read something like this in and be able to output it again in a similar order so it appears familiar.

```as3
var data:Object = JSON.parse( jsonString );

// Then later

trace( JSON.stringify( data, null, 4 ) );
```

results in differing orders, eg:

```json
{
    "name": "My Package",
    "type": "swc",
    "url": "https://example.com/package/",
    "id": "com.my.package",
    "version": "1.0.0",
    "description": "An example of a package definition"
}
```

**Comment by htmiel** [comment, marked as answer]
I needed something similar some time ago (to be able to compare if 2 JSON strings are the same), ended up making a simple class to output a JSON string sorted alphabetically by key names.

But it looks you need to preserve a particular order, which should also be stored somehow in the AS3 "Object"...

**Reply by marchbold** [reply]
Yeah, it seems to output the content in the same order as `for (var key:* in someObject) ... ` loops over the properties in an Object. 

Did you have to completely custom roll your own? 

I might have to use the old as3corelib json encoder and modify it to specify the order. I only need the "top level" order I suppose...

**Reply by htmiel** [reply]
My implementation is pretty basic (compared to the JSONEncoder in as3corelib), but it does the job in my case.
[sortedJSON.txt](https://github.com/airsdk/Adobe-Runtime-Support/files/6707745/sortedJSON.txt)

**Reply by ajwfrost** [reply]
Let us know if you want a tweak to https://github.com/adobe/avmplus/blob/master/core/JSONClass.cpp ...

Looks like if there's a "property list" then things are ordered that way, otherwise it uses the enumarable name indices from the representation of the AS3 object. And the "property list" can be passed in as the second parameter to the `stringify` function (see https://github.com/adobe/avmplus/blob/master/core/JSON.as)

So you would need to know what properties your top-level JSON has, and then pass in an array of these to have it output them in the order defined in the array. Presumably fairly straightforward to do that so that it comes out in alphabetical order...

Curious though, why the initial parsing doesn't put things into the same order. Possibly something to do with the property names that are used:
```
                if (m_toplevel->core()->atomToString(name)->parseIndex(recv)) {
                    o->setUintProperty(recv, val);
                } else {
                    name = m_toplevel->core()->intern(name)->atom();
                    o->setAtomProperty(name, val);
                }
```
The first bit would be for any names that are actually numbers; the second bit puts something into the internal hashtable used for dynamic classes... so it's probably that part that means the write order isn't the same as the read order.

**Reply by ajwfrost** [reply]
Looking at the file from @htmiel  - if you did that initial 'sorted array' thing, you should be able to just pass that into Stringify? for the top level, anyway, it wouldn't sort any objects within the main object...
```
			var IDs:Array = [];
			for (var id:String in sourceObject) {IDs.push(id);}
			IDs.sort();
			JSON.stringify(sourceObject, IDs, 4);
```

**Reply by marchbold** [reply]
Oh, that may work, I thought the replacer had to be a function that was run. 

But passing an array of ordered parameter names would be perfect.  I'll try that in the morning...

**Reply by marchbold** [reply]
@ajwfrost That worked perfectly by the way, I'm passing the ordered keys as a string array and it's outputting JSON in my specified order...  I haven't tried anything more complex, i.e. depth but this is enough for my needs.

**Reply by marchbold** [reply]
Just a note for anyone else using this method, the `IDs` referenced not only filter the top level objects but all keys in the `sourceObject` structure.

**Comment by Adolio** [comment]
Hi @marchbold,

Here is my implementation which is based on Lawrence S. Maccherone, Jr “pretty” JSON encoder which is also based on the as3corelib: https://github.com/Adolio/AS3-Pretty-JSON-Encoder

I used it for several years and it works fine for me.

Best,
Aurélien

**Comment by leossmith** [comment]
`trace( JSON.stringify( data, ['id','name','url','description','type','version'] ) );` should do the trick

**Reply by marchbold** [reply]
Yeah I ended up with a variation on that, had to add all of the "sub" keys as well but this approach is working. eg. had to add `name` and `value` in the below:

```json
{
    "name": "My Package",
    "type": "swc",
    "url": "https://example.com/package/",
    "id": "com.my.package",
    "version": "1.0.0",
    "description": "An example of a package definition",
    "parameters": [
        { 
            "name" : "someconfigurationvalue", 
            "value" : "XXXYYY"
        } 
    ]
}
```

```as3
JSON.stringify( 
    data, 
    ['id','name','url','description','type','version','parameters','name','value'],
    4
);
```

**Reply by leossmith** [reply]
If you are using flex (or import the mx.swc) this will probably also work for you without having to do it manually. _ObjectUtil.getClassInfo(object)_

**Reply by marchbold** [reply]
Yeah I've automated it, definitely not manually adding them ;)
